requirements.txt

```diff
@@ -5,7 +5,8 @@ pycodestyle==2.4.0
 pyflakes==2.0.0
 attrs==18.2.0
 enum34==1.1.6; python_version<"3"
-nose==1.3.7
+pytest==4.6.4
+pytest-cov==2.7.1
```

Comment:
requirements.txt should pin *all* the requirements, including dependencies on pytest. If you create a fresh virtual environment and then `pip install -r requirements.txt` you should be able to determine the dependencies that need pinning.

Reply:
I still don't get why this project has a `requirements.txt` file, which is intended for deployable projects. To create a fresh virtual environment and install all the dependencies `fakeredis` needs, `pip install -e .[testing]` should be used.

As `fakeredis` is a shared library that must work with different Python versions, and also with relaxed dependency versions, having a `requirements.txt` file goes against that purpose.

Please let me know if there's something I'm missing here. Otherwise, I can remove the `requirements.txt` file, and update the documentation to use the `setup.py` file to install requirements (or even better, Tox).

Reply:
It's used for Travis, to ensure that tests are repeatable (and is also handy to locally test in the same environment that is tested on Travis).

Reply:
I can update it in this PR to maintain consistency, but repeatability could be harmful when working on shared libraries. For example, the project requires `sortedcontainers` to be installed, but it only tests `sortedcontainers==2.1.0` because that's what's in `requirements.txt`.

To fix it, we can completely migrate Travis to use Tox environments, and use something like this:
```
[tox]
envlist =
    py{27,34,35,36,37,py}-redis{210,30,31,32,33}

[testenv]
usedevelop = true
commands = pytest -v {posargs}
extras =
    lua
    testing
deps =
    redis210: redis==2.10.6
    redis30: redis==3.0.1
    redis31: redis==3.1.0
    redis32: redis==3.2.1
    redis33: redis==3.3.7
```

Also pinning any other library that could be needed. If tests start to fail because a dependency was updated, it means it's not correctly pinned in the `setup.py` file.

Reply:
It's impractical to test every possible combination of libraries, but at some stage I do want to add another item to the Travis build matrix that will test with the current version of redis, and that could also test with current versions of everything else too. I tend not to use tox much myself so I don't want to try to recreate what I've done with requirements.txt inside tox, and strict pinning in setup.py is not user-friendly as users may have a reason to use a different library version.

Reply:
I have updated this file by using the output from `pip freeze`, and merging versions for Python 2.7 and 3.7. It's clear how impractical it is to maintain this file when the project must work with multiple Python versions, and dependencies need multiple sub-dependencies, but what's how it works today. I can open a separate PR to show how it could work with `setup.py` and Tox.

> strict pinning in setup.py is not user-friendly as users may have a reason to use a different library version.
That's why I specified a `testing` extra, which is only used to test this project, and won't be needed by end users. For them, dependencies will be relaxed, to avoid forcing a particular version on their environments.

Reply:
> I have updated this file by using the output from pip freeze, and merging versions for Python 2.7 and 3.7.

Thanks! I hadn't realised just how much stuff pytest sucks in. I'll merge this now. I'm planning to drop Python 2 support at the end of the year (when Python 2 hits end-of-life) so things will get simpler then.

> I can open a separate PR to show how it could work with setup.py and Tox.

I appreciate the enthusiasm, but I'm happy with the way it works now and how it matches my workflow, so I probably wouldn't merge such a PR.